docs: tw-1008-inside-arbitrum-nitro (#3538)
OffchainLabs/arbitrum-docs: docs/how-arbitrum-works/01-inside-arbitrum-nitro.mdx

## Transaction processing journey on Arbitrum

This document provides a deep dive into the design and rationale of <a data-quicklook-from="arbitrum-nitro">Arbitrum Nitro</a>. This isn't API documentation, nor is it a guided tour of the code—look elsewhere for those. "Inside Arbitrum Nitro" is for people who want to understand Nitro's design.

The majority of this document describes Arbitrum Rollup, the primary use case for Nitro, which runs on the <a data-quicklook-from="arbitrum-one">Arbitrum One</a> chain. There is a variant use case, called <a data-quicklook-from="arbitrum-anytrust-protocol">AnyTrust</a>, which is used by the <a data-quicklook-from="arbitrum-nova">Arbitrum Nova</a> chain. AnyTrust is covered by a section at the end of this document.

<a data-quicklook-from="arbitrum">Arbitrum</a> is an <a data-quicklook-from="L2">L2</a> scaling solution for Ethereum, offering a unique combination of benefits:

> - **Trustless security**: security rooted in Ethereum, with any one party able to ensure correct Layer 2 results.

Reviewer: ```suggestion
- **<a data-quicklook-from="trustless">Trustless</a> security**: security rooted in Ethereum, with any one party able to ensure correct Layer 2 results.
```

> - **Compatibility with Ethereum**: able to run unmodified EVM contracts and Ethereum transactions.

Reviewer: ```suggestion
- **Compatibility with Ethereum**: able to run unmodified EVM contracts and Ethereum <a data-quicklook-from="transaction">transactions</a>.
```

- **Scalability**: moving contracts' computation and storage off of the main Ethereum chain, allowing much higher throughput.
- **Minimum cost**: designed and engineered to minimize the <a data-quicklook-from="L1">L1</a> gas footprint of the system, minimizing per-transaction cost.

Nitro is a major upgrade to Arbitrum, improving over <a data-quicklook-from="arbitrum-classic">"classic" Arbitrum</a> in several ways:

> - **Advanced <a data-quicklook-from="blob">blob</a>/calldata compression** further reduces <a data-quicklook-from="transaction">transaction</a> costs on Arbitrum by reducing the amount of data posted to L1.

Reviewer: ```suggestion
- **Advanced <a data-quicklook-from="blob">blob</a>/calldata compression** further reduces <a data-quicklook-from="transaction">transaction</a> costs on Arbitrum by reducing the amount of data posted to <a data-quicklook-from="L1">L1</a>.
```

- **Separate contexts for common execution and fault proving**, increasing the performance of L1 nodes, and thus offering lower fees.
- **Ethereum L1 gas compatibility**, bringing pricing and accounting for EVM operations perfectly in line with Ethereum.

> - **Additional L1 interoperability**, including tighter synchronization with L1 block numbers, and full support for all Ethereum L1 precompiles.

Reviewer: ```suggestion
- **More L1 interoperability**, including tighter synchronization with L1 block numbers, and full support for all Ethereum L1 precompiles.
```

- **Safe retryables** that eliminate the failure mode where a <a data-quicklook-from="retryable-ticket">retryable ticket</a> fails to be created.
- **<a data-quicklook-from="geth">Geth</a> tracing**, for even broader debugging support.
- And many, many more changes.

## The big picture

At the most basic level, an <a data-quicklook-from="arbitrum-chain">Arbitrum chain</a> works like this:

<ImageWithCaption caption="Original napkin sketch drawn by Arbitrum co-founder Ed Felten" src="/img/haw-eds-napkin-drawing.png" />

People and contracts put messages into the inbox. The chain reads the messages one at a time and processes each one. This updates the chain's state and produces outputs.

> If you want an Arbitrum chain to process a transaction for you, you need to put that transaction into the chain's inbox. Then the chain will see your transaction, execute it, and produce outputs: a transaction receipt and any withdrawals your transaction initiated.

Reviewer: ```suggestion
If you want an <a data-quicklook-from="arbitrum-chain">Arbitrum chain</a> to process a transaction for you, you need to put that transaction into the chain's inbox. Then the chain will see your transaction, execute it, and produce outputs: a transaction receipt and any withdrawals your transaction initiated.
```

Execution is deterministic—which means that the chain's behavior is uniquely determined by the contents of its inbox. Because of this, the result of your transaction is known as soon as it is put in the inbox. Any Arbitrum node can tell you the result. (And you can run an Arbitrum node yourself if you want.)

All the technical details in this document are related to this diagram. To get from this diagram to a full description of Arbitrum, we'll need to answer questions like these:

- Who keeps track of the inbox, <a data-quicklook-from="chain-state">chain state</a>, and outputs?

> - How does Arbitrum make sure that the chain state and outputs are correct?

Reviewer: ```suggestion
- How does Arbitrum make sure that the <a data-quicklook-from="chain-state">chain state</a> and outputs are correct?
```

- How can Ethereum users and contracts interact with Arbitrum?
- How does Arbitrum support Ethereum-compatible contracts and transactions?
- How are **ETH** and tokens transferred into and out of Arbitrum chains, and how are they managed while on the chain?
- How can I run my own Arbitrum node or <a data-quicklook-from="validator">validator</a>?

For foundational concepts about the STF, see the [State Transition Function gentle introduction](/how-arbitrum-works/deep-dives/01-stf-gentle-intro.mdx).

## Nitro's design: the four big ideas

The essence of Nitro, and its key innovations, lie in four big ideas. We'll list them here with a quick summary of each, then we'll unpack them in more detail in later sections.

- **Sequencing, followed by deterministic execution**: Nitro processes transactions with a two-phase strategy. First, the transactions are organized into a single ordered sequence, and Nitro commits to that sequence. Then the transactions are processed, in that sequence, by a deterministic <a data-quicklook-from="state-transition-function">State Transition Function</a>.

> - **Geth at the core**: Nitro supports Ethereum's data structures, formats, and virtual machine by compiling them into the core code of the popular go-ethereum ("Geth") Ethereum node software. Using Geth as a library in this way ensures a very high degree of compatibility with Ethereum.

Reviewer: ```suggestion
- **<a data-quicklook-from="geth">Geth</a> at the core**: Nitro supports Ethereum's data structures, formats, and virtual machine by compiling them into the core code of the popular go-ethereum ("Geth") Ethereum node software. Using Geth as a library in this way ensures a very high degree of compatibility with Ethereum.
```

- **Separating execution from proving**: Nitro compiles the same source code twice, once to native code for execution on a Nitro node, optimized for speed, and again to <a data-quicklook-from="wasm">WASM</a> for proving, optimized for portability and security.

> - **Optimistic Rollup with interactive fraud proofs**: Nitro settles transactions on the Layer 1 Ethereum chain using an Optimistic Rollup protocol that includes interactive fraud proofs, pioneered by Arbitrum. The current dispute system, [BoLD](/how-arbitrum-works/bold/gentle-introduction.mdx), extends this with an all-vs-all, <a data-quicklook-from="permissionless-validation">permissionless validation</a> protocol that resolves disputes with a bounded delay (See "Resolving disputes using interactive fraud proofs").

Reviewer: ```suggestion
- **Optimistic Rollup with interactive <a data-quicklook-from="fraud-proof">fraud proofs</a>**: Nitro settles transactions on the Layer 1 Ethereum chain using an Optimistic Rollup protocol that includes interactive fraud proofs, pioneered by Arbitrum. The current dispute system, [BoLD](/how-arbitrum-works/bold/gentle-introduction.mdx), extends this with an all-vs-all, <a data-quicklook-from="permissionless-validation">permissionless validation</a> protocol that resolves disputes with a bounded delay (See "Resolving disputes using interactive fraud proofs").
```

## Sequencing, followed by deterministic execution

This diagram summarizes how transactions are processed in Nitro.

<ImageZoom src="/img/haw-transaction-lifecycle.svg" className="img-800px" alt="Transaction lifecycle">
  Transaction lifecycle
</ImageZoom>

Let's [follow a user's transaction](/how-arbitrum-works/deep-dives/transaction-lifecycle.mdx) through this process.

> First, the user creates a transaction, signs it with their <a data-quicklook-from="wallet">wallet</a>, and sends it to the Nitro chain's [Sequencer](/how-arbitrum-works/deep-dives/sequencer.mdx). The Sequencer's job, as its name implies, is to take the arriving transactions, put them into an ordered sequence, and publish that sequence.

Reviewer: ```suggestion
First, the user creates a transaction, signs it with their <a data-quicklook-from="wallet">wallet</a>, and sends it to the Nitro chain's [Sequencer](/how-arbitrum-works/deep-dives/sequencer.mdx). The <a data-quicklook-from="sequencer">Sequencer</a>'s job, as its name implies, is to take the arriving transactions, put them into an ordered sequence, and publish that sequence.
```

> Once the transactions are sequenced, they are run through the _state transition function_, one by one. The state transition function (STF) takes as input the current state of the chain (account balances, contract code, and so on) along with the next transaction. It updates the state and sometimes emits a new Layer 2 block on the Nitro chain.

Reviewer: ```suggestion
Once the transactions are sequenced, they are run through the _<a data-quicklook-from="state-transition-function">state transition function</a>_, one by one. The state transition function (STF) takes as input the current state of the chain (account balances, contract code, and so on) along with the next transaction. It updates the state and sometimes emits a new Layer 2 block on the Nitro chain.
```

> Because the protocol doesn't trust the <a data-quicklook-from="sequencer">Sequencer</a> not to put garbage into its sequence, the STF will detect and discard invalid (e.g., improperly formed) transactions from the sequence. A well-behaved Sequencer will filter out invalid transactions so the STF never sees them—and this reduces cost and therefore keeps transaction fees low—but Nitro will still work correctly no matter what the Sequencer puts into its feed. (Transactions in the feed are signed by their senders, so the Sequencer can't create forged transactions.)

Reviewer: ```suggestion
Because the protocol doesn't trust the <a data-quicklook-from="sequencer">Sequencer</a> not to put garbage into its sequence, the STF will detect and discard invalid (for example, improperly formed) transactions from the sequence. A well-behaved Sequencer will filter out invalid transactions so the STF never sees them—and this reduces cost and therefore keeps transaction fees low—but Nitro will still work correctly no matter what the Sequencer puts into its feed. (Transactions in the feed are signed by their senders, so the Sequencer can't create forged transactions.)
```

The State Transition Function is deterministic, which means that its behavior depends only on the current state and the contents of the next transaction—and nothing else. Because of this determinism, the result of a transaction `T` will depend only on the genesis state of the chain, the transactions before `T` in the sequence, and `T` itself.

It follows that anyone who knows the transaction sequence can compute the STF themselves, and all honest parties who do so will receive identical results. This is the normal way that Nitro nodes operate: get the transaction sequence and run the State Transition Function locally. No consensus mechanism is needed for this.

## How the Sequencer publishes the sequence

So how do nodes get the sequence? The Sequencer publishes it in two ways:

- a <a data-quicklook-from="sequencer-feed">real-time feed</a>, and

> - batches posted on L1 Ethereum

Reviewer: ```suggestion
- <a data-quicklook-from="batch">batches</a> posted on L1 Ethereum
```

The real-time feed is published by the Sequencer, so anyone who subscribes receives instant notifications of each transaction as it is sequenced. Nitro nodes can subscribe to the feed directly from the Sequencer or via a <a data-quicklook-from="feed-relay">relay</a> that forwards it. The feed represents the Sequencer's promise to record transactions in a particular order. If the Sequencer is honest and doesn't have a long downtime, this promise will be kept. So anyone who trusts the Sequencer to keep its promises can rely on the feed to get instant information about the transaction sequence—and they can run the sequenced transactions through the State Transition Function to learn the results of each transaction immediately. This is "<a data-quicklook-from="soft-confirmation">soft finality</a>" for transactions; it's "soft" because it depends on the Sequencer keeping its promises.

> The Sequencer also publishes its sequence on the L1 Ethereum chain. Periodically—perhaps every few minutes in production—the Sequencer concatenates the next group of transactions in the feed, compresses them for efficiency, and posts the results to Ethereum—by default as EIP-4844 blobs when the <a data-quicklook-from="parent-chain">parent chain</a> supports them, and they are priced efficiently—otherwise as calldata. This is the final and official record of the transaction sequence. As soon as this Ethereum transaction reaches finality, the Layer 2 Nitro transactions it records will also have finality. These transactions are final because their positions in the sequence are final, and their outcomes are deterministic and knowable to any party. This is "<a data-quicklook-from="hard-finality">hard finality</a>."

Reviewer: ```suggestion
The Sequencer also publishes its sequence on the L1 Ethereum chain. Periodically—perhaps every few minutes in production—the Sequencer concatenates the next group of transactions in the feed, compresses them for efficiency, and posts the results to Ethereum—by default as EIP-4844 <a data-quicklook-from="blob">blobs</a> when the <a data-quicklook-from="parent-chain">parent chain</a> supports them, and they are priced efficiently—otherwise as calldata. This is the final and official record of the transaction sequence. As soon as this Ethereum transaction reaches finality, the Layer 2 Nitro transactions it records will also have finality. These transactions are final because their positions in the sequence are final, and their outcomes are deterministic and knowable to any party. This is "<a data-quicklook-from="hard-finality">hard finality</a>."
```

> The Sequencer's <a data-quicklook-from="batch">batches</a> are compressed using the general-purpose Brotli data compression algorithm. Rather than always using the maximum compression level, the batch poster selects a Brotli level adaptively based on its backlog (the default at low backlog is the maximum level 11, dropping to faster, lower levels as the backlog grows).

Reviewer: ```suggestion
The Sequencer's <a data-quicklook-from="batch">batches</a> are compressed using the general-purpose Brotli data compression algorithm. Rather than always using the maximum compression level, the <a data-quicklook-from="batch-poster">batch poster</a> selects a Brotli level adaptively based on its backlog (the default at low backlog is the maximum level 11, dropping to faster, lower levels as the backlog grows).
```

## Geth at the core

The second key design idea in Nitro is "Geth at the core." Here, "Geth" refers to [go-ethereum](/how-arbitrum-works/reference/geth.mdx), the most common node software for Ethereum. As its name would suggest, go-ethereum is written in the Go programming language, as is almost all of Nitro.

> <VanillaAdmonition type="info">In Ethereum's post-Merge architecture, Geth is an **execution client**: it runs the EVM and maintains state, working alongside a separate consensus client. Nitro doesn't run Geth as a standalone client—it compiles in Geth's execution core as a library (the "Geth sandwich" described below). Nitro also has no consensus client; the Sequencer and the Layer 1 Rollup protocol handle sequencing and finality instead.</VanillaAdmonition>

Reviewer: ```suggestion
<VanillaAdmonition type="info">In Ethereum's post-Merge architecture, Geth is an **execution <a data-quicklook-from="client">client</a>**: it runs the EVM and maintains state, working alongside a separate consensus client. Nitro doesn't run Geth as a standalone client—it compiles in Geth's execution core as a library (the "Geth sandwich" described below). Nitro also has no consensus client; the Sequencer and the Layer 1 Rollup protocol handle sequencing and finality instead.</VanillaAdmonition>
```

- The base layer is the core of Geth—the parts that emulate the execution of EVM contracts and maintain the data structures that comprise the Ethereum state. Nitro compiles in this code as a library, with a few minor modifications to add necessary hooks.
- The middle layer, which we call [ArbOS](/how-arbitrum-works/deep-dives/arbos.mdx), is custom software that provides additional functions associated with Layer 2 functionality, such as decompressing and parsing the Sequencer's data batches, accounting for Layer 1 gas costs and collecting fees to reimburse for them, and supporting cross-chain <a data-quicklook-from="bridge">bridge</a> functionalities such as deposits of Ether (**ETH**) and tokens from L1 and withdrawals of the same back to L1. We'll dig into the details of <a data-quicklook-from="arbos">ArbOS</a> below.
- The top layer consists of node software, mostly drawn from Geth. This handles connections and incoming RPC requests from clients and provides the other top-level functionality required to operate an Ethereum-compatible <a data-quicklook-from="blockchain">blockchain</a> node.

> Because the top and bottom layers rely heavily on code from Geth, this structure has been dubbed a "Geth sandwich." Strictly speaking, Geth plays the role of the bread in the sandwich, and ArbOS is the filling, but this sandwich is named for the bread.

Reviewer: ```suggestion
Because the top and bottom layers rely heavily on code from Geth, this structure has been dubbed a "Geth sandwich." Strictly speaking, Geth plays the role of the bread in the sandwich, and <a data-quicklook-from="arbos">ArbOS</a> is the filling, but this sandwich is named for the bread.
```

The State Transition Function consists of the bottom Geth layer and a portion of the middle ArbOS layer. In particular, the STF is a designated function in the source code and implicitly includes all code that it calls. The STF takes as input the [bytes of a transaction received in the inbox](/how-arbitrum-works/reference/stf-inputs.mdx) and has access to a modifiable copy of the Ethereum state tree. Executing the STF may modify the state and, upon completion, emit the header of a new block (in Ethereum's block header format), which will be appended to the Nitro chain.

## Separating execution from proving

> One of the challenges in designing a practical rollup system is the tension between wanting the system to perform well in ordinary execution and reliably verifying its results. Nitro resolves this tension by using the same source code for both execution and proving, but compiling it to different targets for the two cases.

Reviewer: ```suggestion
One of the <a data-quicklook-from="challenge">challenges</a> in designing a practical rollup system is the tension between wanting the system to perform well in ordinary execution and reliably verifying its results. Nitro resolves this tension by using the same source code for both execution and proving, but compiling it to different targets for the two cases.
```

> When compiling the Nitro node software for _execution_, the standard Go compiler produces native code for the target architecture, which, of course, will vary across node deployments. (The node software is distributed in source code form, and as a Docker image containing a compiled library.)

Reviewer: ```suggestion
When compiling the Nitro node software for _execution_, the standard Go compiler produces native code for the target architecture, which will vary across node deployments. (The node software is distributed in source code form, and as a Docker image containing a compiled library.)
```

> Separately, for proof, the State Transition Function portion of the code is compiled to WebAssembly (WASM), a typed, portable machine code format. The WASM code is then transformed into a format we call <a data-quicklook-from="wavm">WAVM</a>, as detailed below. If there is a dispute over the correct result of the STF computation, it is resolved to the WAVM code.

Reviewer: ```suggestion
Separately, for proof, the State Transition Function portion of the code is compiled to WebAssembly (<a data-quicklook-from="wasm">WASM</a>), a typed, portable machine code format. The WASM code is then transformed into a format we call <a data-quicklook-from="wavm">WAVM</a>, as detailed below. If there is a dispute over the correct result of the STF computation, it is resolved to the WAVM code.
```

The "hash <a data-quicklook-from="oracle">oracle</a> trick" of storing the Merkle hash of a data structure and relying on protocol participants to store the full structure to support fetch-by-hash of its contents dates back to the original Arbitrum design.

## Optimistic Rollup

Arbitrum is an Optimistic Rollup. Let's unpack what that means.

#### Rollup

Arbitrum is a Rollup, which means that the inputs to the chain—the messages that are put into the inbox—are all recorded on the Ethereum chain as EIP-4844 blobs or as calldata if blobs are unavailable. Because of this, everyone has the information they need to determine the chain's current state—they have the full history of the inbox, and the results are uniquely determined by that history, so they can reconstruct the chain's state based on public information, if needed.

> This also allows anyone to participate in the Arbitrum protocol by running an Arbitrum node or serving as a validator. Nothing about the chain's history or state is a secret.

Reviewer: ```suggestion
This also allows anyone to participate in the Arbitrum protocol by running an Arbitrum node or serving as a <a data-quicklook-from="validator">validator</a>. Nothing about the chain's history or state is a secret.
```

#### Optimistic

> Arbitrum is optimistic, meaning that it advances the state of the chain by allowing any party (a "validator") to post an <a data-quicklook-from="assertion">assertion</a> on Layer 1 that the party claims is correct, and then giving everyone else a chance to challenge that claim. If the <a data-quicklook-from="challenge-period">challenge period</a> (6.4 days) passes and no one challenges the claimed assertion, Arbitrum confirms it as correct. If someone challenges the claim during the challenge period, Arbitrum uses an efficient dispute-resolution protocol (detailed below) to determine which party is lying. The liar will forfeit a deposit, and the truth-teller will take part of that deposit as a reward for their efforts (some of the deposit is burned, guaranteeing that the liar is punished even if there's some collusion going on).

Reviewer: ```suggestion
Arbitrum is optimistic, meaning that it advances the state of the chain by allowing any party (a "validator") to post an <a data-quicklook-from="assertion">assertion</a> on Layer 1 that the party claims is correct, and then giving everyone else a chance to challenge that claim. If the <a data-quicklook-from="challenge-period">challenge period</a> (6.4 days) passes and no one challenges the claimed <a data-quicklook-from="assertion">assertion</a>, Arbitrum confirms it as correct. If someone challenges the claim during the challenge period, Arbitrum uses an efficient dispute-resolution protocol (detailed below) to determine which party is lying. The liar will forfeit a deposit, and the truth-teller will take part of that deposit as a reward for their efforts (some of the deposit is burned, guaranteeing that the liar is punished even if there's some collusion going on).
```

### Why interactive proving is better

We strongly believe that interactive proving is the superior approach, for the following reasons:

- **More efficient in the optimistic case**: Because interactive proving can resolve disputes that are larger than one transaction, it can allow an assertion to contain only a single claim about the end state of the chain after all of the execution covered by the assertion. By contrast, re-executing requires posting a state claim for each transaction within the assertion. With hundreds or thousands of transactions per assertion, this is a substantial difference in L1 footprint—and L1 footprint is the main component of cost.
- **More efficient in the pessimistic case**: In the case of a dispute, interactive proving requires the L1 mediator contract only to check that Alice and Bob's actions "have the right shape," for example, that Alice has divided her `N`-step claim into two claims half as large. (The mediator doesn't need to evaluate the correctness of Alice's claim—Bob does that, offchain.) Only one instruction needs to be re-executed. By contrast, re-execution requires the L1 mediator to emulate the execution of an entire transaction.

> - **High per-transaction gas limit**: Interactive proving can escape from Ethereum's tight per-transaction gas limit. The gas limit isn't infinite, for obvious reasons, but it can be larger than on Ethereum. As far as Ethereum is concerned, the only downside of a gas-heavy Arbitrum transaction is that it may require an interactive <a data-quicklook-from="fraud-proof">fraud proof</a> with slightly more steps (and only if indeed it is fraudulent). By contrast, re-execution must impose a _lower_ gas limit than Ethereum, because it must be possible to emulate execution of the transaction (which is more expensive than executing it directly) within a single Ethereum transaction.

Reviewer: ```suggestion
- **High per-transaction gas limit**: Interactive proving can escape from Ethereum's tight per-transaction gas limit. The gas limit isn't infinite, for practical reasons, but it can be larger than on Ethereum. As far as Ethereum is concerned, the only downside of a gas-heavy Arbitrum transaction is that it may require an interactive <a data-quicklook-from="fraud-proof">fraud proof</a> with slightly more steps (and only if indeed it is fraudulent). By contrast, re-execution must impose a _lower_ gas limit than Ethereum, because it must be possible to emulate execution of the transaction (which is more expensive than executing it directly) within a single Ethereum transaction.
```

With those preliminaries behind us, let's jump into the details of the rollup protocol.

The parties that participate in the protocol are called _validators_. Some validators choose to be <a data-quicklook-from="bonder">bonders</a>—they place a deposit that they can recover if they're not caught cheating. In the common case, it's expected that only one validator will be bonded, since as long as it's staked on the current outcome, and there are no conflicting claims, there's no need for other parties to bond or take any action.

With the deployment of <a data-quicklook-from="bold">BoLD</a>, validation is permissionless: anyone can post an assertion, bond, and defend the chain. (A chain may still keep an optional validator allowlist enabled as a safety measure during early stages; this is a toggle in the rollup contracts, not a requirement of the protocol.) "<a data-quicklook-from="watchtower-validator">Watchtower validators</a>," who monitor the chain but don't take any onchain actions, can be run permissionlessly (see "validators" below).

> The key security property of the rollup protocol is that any one honest validator can force the correct execution of the chain to be confirmed. This means that executing an Arbitrum chain is as trustless as executing an Ethereum chain. You, and you alone (or someone you hire), can ensure your transactions are processed correctly. And this is true no matter how many malicious people are trying to stop you. Under BoLD, this guarantee comes with a [_bounded delay_](/how-arbitrum-works/bold/bold-technical-deep-dive.mdx): a single honest party can force <a data-quicklook-from="confirmation">confirmation</a> of the correct outcome with a known, fixed amount of time, regardless of how many adversaries participate or how much they place as bond.

Reviewer: ```suggestion
The key security property of the rollup protocol is that any one honest validator can force the correct execution of the chain to be confirmed. This means that executing an Arbitrum chain is as trustless as executing an Ethereum chain. You, and you alone (or someone you hire), can ensure your transactions are processed correctly. And this is true no matter how many malicious people are trying to stop you. Under <a data-quicklook-from="bold">BoLD</a>, this guarantee comes with a [_bounded delay_](/how-arbitrum-works/bold/bold-technical-deep-dive.mdx): a single honest party can force <a data-quicklook-from="confirmation">confirmation</a> of the correct outcome with a known, fixed amount of time, regardless of how many adversaries participate or how much they place as bond.
```

### The Rollup chain

> The rollup protocol tracks a chain of claims about the L2 state—we'll call these [assertions](/how-arbitrum-works/deep-dives/assertions.mdx). They're not the same as Layer 1 Ethereum blocks, and also not the same as Layer 2 (Nitro) blocks. You can think of these assertions as forming a tree of claims about the chain's history, which the <a data-quicklook-from="arbitrum-rollup-protocol">Arbitrum Rollup Protocol</a> manages and oversees.

Reviewer: ```suggestion
The rollup protocol tracks a chain of claims about the <a data-quicklook-from="L2">L2</a> state—we'll call these [assertions](/how-arbitrum-works/deep-dives/assertions.mdx). They're not the same as Layer 1 Ethereum blocks, and also not the same as Layer 2 (Nitro) blocks. You can think of these assertions as forming a tree of claims about the chain's history, which the <a data-quicklook-from="arbitrum-rollup-protocol">Arbitrum Rollup Protocol</a> manages and oversees.
```

Validators can propose assertions. A new assertion is initially unresolved. Eventually, every assertion will be _resolved_ by being either _confirmed_ or _rejected_. The confirmed assertions constitute the chain's confirmed history.

Each assertion contains, among other things:

- The <a data-quicklook-from="predecessor-assertion">predecessor assertion</a> it builds on (the last assertion before this one that is claimed to be correct)

> - The number of L2 blocks that have been created in the chain's history

Reviewer: ```suggestion
- The number of <a data-quicklook-from="l2-block">L2 blocks</a> that have been created in the chain's history
```

- The number of inbox messages that have been consumed in the chain's history

> - A commitment to the state and outputs produced over the chain's history

Reviewer: ```suggestion
- A <a data-quicklook-from="commitment">commitment</a> to the state and outputs produced over the chain's history
```

Except for its position in the tree, the contents of an assertion are all just claims by its <a data-quicklook-from="proposer">proposer</a>. Arbitrum doesn't know at first whether they are correct. If they are correct, the protocol should eventually confirm the assertion. If they are incorrect, the protocol should eventually reject them.

An assertion implicitly claims that its predecessor is correct, and therefore, transitively, that a complete history of the chain reaching back to its genesis is correct. It also implicitly claims that any <a data-quicklook-from="rival">rival</a> assertion—a different assertion sharing the same predecessor—is incorrect.

> In the normal case, only correct assertions are proposed. If an assertion has no rival and its challenge period elapses, it is confirmed. If two or more rival assertions exist, they are resolved by a <a data-quicklook-from="challenge">challenge</a> (described below), and the assertions descending from the losing side are rejected. The crucial property of BoLD is that this resolution completes with a bounded delay, no matter how many rival assertions a malicious party creates.

Reviewer: ```suggestion
In the normal case, only correct assertions are proposed. If an assertion has no <a data-quicklook-from="rival">rival</a> and its challenge period elapses, it is confirmed. If two or more rival assertions exist, they are resolved by a <a data-quicklook-from="challenge">challenge</a> (described below), and the assertions descending from the losing side are rejected. The crucial property of BoLD is that this resolution completes with a bounded delay, no matter how many rival assertions a malicious party creates.
```

### Bonding

> At any given time, some validators will be bonders, and some will not. To propose an assertion, a validator must put up a [bond held by the Arbitrum Layer 1 contracts](/how-arbitrum-works/bold/bold-economics-of-disputes.mdx), which will be confiscated if the assertion is proven false. The bond is denominated in an **ERC-20** token configured for the chain—on Arbitrum One, this is **WETH**—so, in practice, stakers post **ETH**-denominated value.

Reviewer: ```suggestion
At any given time, some validators will be <a data-quicklook-from="bonder">bonders</a>, and some will not. To propose an assertion, a validator must put up a [bond held by the Arbitrum Layer 1 contracts](/how-arbitrum-works/bold/bold-economics-of-disputes.mdx), which will be confiscated if the assertion is proven false. The bond is denominated in an **ERC-20** token configured for the chain—on Arbitrum One, this is **WETH**—so, in practice, stakers post **ETH**-denominated value.
```

A validator who agrees with the current confirmed history and proposes a correct successor assertion stays bonded on it until it is confirmed, at which point the bond can be withdrawn. A validator whose assertion is rejected forfeits its bond.

> When a dispute arises, the challenge game itself is played over _edges_ (see "Challenges"), and creating an <a data-quicklook-from="edge">edge</a> requires posting a small additional _mini-bond_ at each level of the dispute. Because only one assertion in a set of rivals can ever be confirmed, the protocol only needs to retain a single mini-bond per level; the redundant bonds posted by rivals can be released to a designated excess-bond receiver.

Reviewer: ```suggestion
When a dispute arises, the challenge game itself is played over _<a data-quicklook-from="edge">edges</a>_ (see "Challenges"), and creating an <a data-quicklook-from="edge">edge</a> requires posting a small additional _mini-bond_ at each level of the dispute. Because only one assertion in a set of rivals can ever be confirmed, the protocol only needs to retain a single mini-bond per level; the redundant bonds posted by rivals can be released to a designated excess-bond receiver.
```

> Under BoLD, the bond required to propose an assertion and the mini-bonds required to participate in a challenge are fixed parameters of the chain (the per-level mini-bond amounts are configured in the challenge manager). There is no exponential escalation tied to confirmation delay, because BoLD's bounded-delay guarantee removes the delay-attack incentive that the escalation was designed to counter: an adversary can't push out confirmation by piling on false stakes, so there is no need to raise the price of doing so over time.

Reviewer: ```suggestion
Under BoLD, the bond required to propose an assertion and the mini-bonds required to participate in a challenge are fixed parameters of the chain (the per-level mini-bond amounts are configured in the challenge manager). There is no exponential escalation tied to <a data-quicklook-from="confirmation">confirmation</a> delay, because BoLD's bounded-delay guarantee removes the delay-attack incentive that the escalation was designed to counter: an adversary can't push out confirmation by piling on false stakes, so there is no need to raise the price of doing so over time.
```

### Rules for confirming or rejecting

The rules for resolving assertions are straightforward.

An assertion can be confirmed if:

- Its predecessor is the latest confirmed assertion, and
- Its challenge period has elapsed, and
- It has no rival (no competing assertions share its predecessor), **or** the challenge among the rivals has been resolved in its favor via the edge protocol

An assertion is rejected if its predecessor is rejected, or if a rival assertion wins the challenge between them.

> A consequence of these rules is that a lone, unrivaled assertion is simply confirmed once its time elapses—no challenge is ever needed in the common, honest case. When two or more rival assertions exist, the disagreement is settled by a challenge. It's time to look at how those work.

Reviewer: ```suggestion
A consequence of these rules is that a lone, unrivaled assertion is confirmed once its time elapses—no challenge is ever needed in the common, honest case. When two or more rival assertions exist, the disagreement is settled by a challenge. It's time to look at how those work.
```

### Challenges

Suppose two (or more) rival assertions have been posted, sharing the same predecessor. They make conflicting claims about the chain's history, so at most one of them can be correct.

> Whenever rival assertions exist, anyone—not just the proposers—can open a challenge. The rollup protocol records the challenge and mediates it via the **Edge Challenge Manager**, eventually confirming the correct side and allowing the losing side's bond to be confiscated.

Reviewer: ```suggestion
Whenever rival assertions exist, anyone—not just the <a data-quicklook-from="proposer">proposers</a>—can open a challenge. The rollup protocol records the challenge and mediates it via the **Edge Challenge Manager**, eventually confirming the correct side and allowing the losing side's bond to be confiscated.
```

<a data-quicklook-from="dissection">Dissection</a> proceeds across several levels—first over Layer 2 blocks, then over progressively finer ranges of WAVM instructions (a block level, one or more big-step levels, and a small-step level)—until an edge represents a single step of execution. The final edge is confirmed by a **one-step proof**, checked onchain by the one-step proof entry contract, which determines who was telling the truth about that one instruction.

> The two subsections below give the classic two-party intuition for _why_ dissection works. They describe a simplified Alice-vs-Bob game; in BoLD, the same logic applies, generalized so that any number of parties can compete and the honest party always has a winning line of play.

Reviewer: ```suggestion
The two subsections below give the classic two-party intuition for _why_ <a data-quicklook-from="dissection">dissection</a> works. They describe a simplified Alice-vs-Bob game; in BoLD, the same logic applies, generalized so that any number of parties can compete and the honest party always has a winning line of play.
```

#### Dissection protocol: simplified version

> Alice is defending the claim that, starting from the state in the predecessor assertion, the Virtual Machine can advance to the state specified in her assertion. Essentially, she is claiming that the Virtual Machine can execute `N` instructions, and that the execution will consume `M` inbox messages and transform the hash of outputs from `H'` to `H`.

Reviewer: ```suggestion
Alice is defending the claim that, starting from the state in the <a data-quicklook-from="predecessor-assertion">predecessor assertion</a>, the Virtual Machine can advance to the state specified in her assertion. Essentially, she is claiming that the Virtual Machine can execute `N` instructions, and that the execution will consume `M` inbox messages and transform the hash of outputs from `H'` to `H`.
```

(If you're a stickler for mathematical precision, it should be clear how these arguments can be turned into proofs by induction on `N`.)

#### The real decision protocol

The real dissection protocol is conceptually similar to the simplified one described above, but with several changes that improve efficiency, enable permissionless all-vs-all play, and deal with necessary corner cases. Here is a list of the differences:

- **Dissection over L2 blocks, then over instructions**: An assertion asserts the result of creating some number of Layer 2 Nitro blocks. Dissection first occurs over these Layer 2 blocks to narrow the dispute down to a single Layer 2 Nitro block. At that point, the dispute becomes about a single execution of the State Transition Function—i.e., the execution of a sequence of WAVM instructions—and the recursive dissection sub-protocol runs again over WAVM instructions (across a block level, one or more big-step levels, and a small-step level) to narrow the dispute to a single instruction. The dispute concludes with a one-step proof of a single instruction (or a party failing to act and losing on time).

> - **Binary <a data-quicklook-from="bisection">bisection</a>**: Each edge is split into two half-size child edges by posting a single midpoint claim. (The classic protocol generalizes this to K-way dissection; BoLD uses straightforward binary bisection).

Reviewer: ```suggestion
- **Binary <a data-quicklook-from="bisection">bisection</a>**: Each edge is split into two half-size child edges by posting a single midpoint claim. (The classic protocol generalizes this to K-way dissection; BoLD uses straightforward binary <a data-quicklook-from="bisection">bisection</a>).
```

- **Deal with the Empty-Inbox case**: The machine can't always execute `N` steps without getting stuck. It might halt, or it might have to wait because its inbox is exhausted, so it can't go on until more messages arrive. The protocol, therefore, allows a challenger to claim that the specified amount of execution is not possible under the current conditions, rather than being forced to assert a concrete end state.
- **Time limits**: Confirmation is governed by per-edge, unrivaled timers measured against a challenge period (6.4 days), and the protocol limits the total time to confirmation. A participant who stops responding while behind will lose once the relevant time elapses.

> It should be clear that these changes don't affect the basic correctness of the challenge protocol. They do, however, improve their efficiency, make validation permissionless, and bound the worst-case delay.

Reviewer: ```suggestion
It should be clear that these changes don't affect the basic correctness of the <a data-quicklook-from="challenge-protocol">challenge protocol</a>. They do, however, improve their efficiency, make validation permissionless, and bound the worst-case delay.
```
